libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/LUIS/LuisAdaptiveRecognizer.cs

```diff
@@ -0,0 +1,131 @@
+        [JsonProperty("endpointKey")]
+        public StringExpression EndpointKey { get; set; }
+
+        /// <summary>
+        /// Gets or sets an external entity recognizer.
+        /// </summary>
+        /// <remarks>This recognizer is run before calling LUIS and the results are passed to LUIS.</remarks>
+        /// <value>Recognizer.</value>
+        [JsonProperty("externalEntityRecognizer")]
+        public Recognizer ExternalEntityRecognizer { get; set; }
+
+        /// <summary>
+        /// Gets or sets an expression or constant LUIS dynamic list.
+        /// </summary>
+        /// <value>Dynamic lists.</value>
+        [JsonProperty("dynamicLists")]
+        public ArrayExpression<Luis.DynamicList> DynamicLists { get; set; }
+
+        /// <summary>
+        /// Gets or sets LUIS prediction options.
+        /// </summary>
+        /// <value>Prediction options.</value>
+        [JsonProperty("predictionOptions")]
+        public AI.LuisV3.LuisPredictionOptions PredictionOptions { get; set; } = new AI.LuisV3.LuisPredictionOptions();
+
+        /// <summary>
+        /// Gets or sets HTTP client handler.
+        /// </summary>
+        /// <value>HTTP client handler.</value>
+        [JsonIgnore]
+        public HttpClientHandler HttpClient { get; set; }
+
+        /// <summary>
+        /// Gets or sets <see cref="IBotTelemetryClient"/>.
+        /// </summary>
+        /// <value>Telemetry client.</value>
+        [JsonIgnore]
+        public IBotTelemetryClient TelemetryClient { get; set; } = new NullBotTelemetryClient();
+
+        /// <inheritdoc/>
+        public override async Task<RecognizerResult> RecognizeAsync(DialogContext dialogContext, string text, string locale, CancellationToken cancellationToken = default)
+        {
+            var wrapper = new LuisRecognizer(RecognizerOptions(dialogContext), HttpClient);
+            var context = dialogContext.Context;
+            if (context.Activity == null || context.Activity.Type != ActivityTypes.Message || context.Activity.Text != text || context.Activity.Locale != locale)
+            {
+                throw new ArgumentException("TurnContext is different than text");
+            }
+
+            return await RecognizeAsync(dialogContext, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Construct V3 recognizer options from the current dialog context.
+        /// </summary>
+        /// <param name="dialogContext">Context.</param>
+        /// <returns>LUIS Recognizer options.</returns>
+        public LuisRecognizerOptionsV3 RecognizerOptions(DialogContext dialogContext)
+        {
+            var options = PredictionOptions;
+            if (DynamicLists != null)
+            {
+                options = new AI.LuisV3.LuisPredictionOptions(options);
+                var list = new List<AI.LuisV3.DynamicList>();
+                foreach (var listEntity in DynamicLists.GetValue(dialogContext.GetState()))
```

Comment:
>dialogContext.GetState() [](start = 65, length = 24)

this allocates a class internally, it's better to assign to dcState and use that in the loop.

Reply:
Umm.  This is only evaluated once to start the loop by getting the value.